cudax/include/cuda/experimental/__container/vector.cuh

```diff
@@ -0,0 +1,1585 @@
+  }
+
+  //! @brief Value-initializes elements in the range `[__first, __first + __count)`.
+  //! @param __first Pointer to the first element to be initialized.
+  //! @param __count The number of elements to be initialized.
+  static void __value_initialize_n(pointer __first, size_type __count)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      _CUDA_VSTD::fill_n(__first, __count, _Tp());
+    }
+    else
+    {
+      thrust::fill_n(thrust::device, __first, __count, _Tp());
+    }
+  }
+
+  //! @brief Copy-constructs elements in the range `[__first, __first + __count)`.
+  //! @param __first Pointer to the first element to be initialized.
+  //! @param __count The number of elements to be initialized.
+  static void __fill_n(pointer __first, size_type __count, const _Tp& __value)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      _CUDA_VSTD::fill_n(__first, __count, __value);
+    }
+    else
+    {
+      thrust::fill_n(thrust::device, __first, __count, __value);
+    }
+  }
+
+  //! @brief Removes elements within the sequence `[__first, __last)` that are equal to \p __value.
+  //! @param __first Pointer to the start of the sequence.
+  //! @param __last Pointer to the end of the sequence.
+  //! @param __value The element to be removed.
+  static pointer __remove_value(pointer __first, pointer __last, const _Tp& __value)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::remove(__first, __last, __value);
+    }
+    else
+    {
+      return thrust::remove(thrust::device, __first, __last, __value);
+    }
+    _CCCL_UNREACHABLE();
+  }
+
+  //! @brief Removes elements within the sequence `[__first, __last)` that are satisfy \p __pred.
+  //! @param __first Pointer to the start of the sequence.
+  //! @param __last Pointer to the end of the sequence.
+  //! @param __pred The predicate to select elements for removal.
+  template <class _Predicate>
+  static pointer __remove_pred(pointer __first, pointer __last, _Predicate __pred)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::remove_if(__first, __last, __pred);
+    }
+    else
+    {
+      return thrust::remove_if(thrust::device, __first, __last, __pred);
+    }
+    _CCCL_UNREACHABLE();
+  }
+
+  //! @brief Equality-compares elements from the sequence `[__first1, __last1)` with those of sequence
+  //! `[__first2, __last2)`.
+  //! @param __first1 Pointer to the start of the first sequence.
+  //! @param __last1 Pointer to the end of the first sequence.
+  //! @param __first2 Pointer to the start of the second sequence.
+  //! @param __last2 Pointer to the end of the second sequence.
+  static bool __equality(const_pointer __first1, const_pointer __last1, const_pointer __first2, const_pointer __last2)
+  {
+    _CCCL_IF_CONSTEXPR (__is_host_only)
+    {
+      return _CUDA_VSTD::equal(__first1, __last1, __first2, __last2);
+    }
+    else
+    {
+      return ((__last1 - __first1) == (__last2 - __first2))
```

Comment:
why doesn't `thrust::equal` implement this shortcut when given random-access iterators?

Reply:
Because thrust only takes `first2` and not both `first2` and `last2`